File: client/net/client/Com.Bj58.Spat.Gaea.Serializer/Component/TypeMap.cs
/*
 *  Copyright Beijing 58 Information Technology Co.,Ltd.
 *
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
//----------------------------------------------------------------------
//<Copyright company="58.com">
//      Team:SPAT
//      Blog:http://blog.58.com/spat/  
//      Website:http://www.58.com
//</Copyright>
//-----------------------------------------------------------------------
using System;
using System.IO;
using System.Linq;
using System.Text;
using System.Reflection;
using System.Collections;
using System.Collections.Generic;
using Com.Bj58.Spat.Gaea.Serializer;
using Com.Bj58.Spat.Gaea.Serializer.Component.Attributes;
using Com.Bj58.Spat.Gaea.Serializer.Component.Exceptions;

namespace Com.Bj58.Spat.Gaea.Serializer.Component
{
    public static class TypeMap
    {
        private static Dictionary<Type, ClassItem> TypeIdMap = new Dictionary<Type, ClassItem>();
        private static Dictionary<uint, ClassItem> IdTypeMap = new Dictionary<uint, ClassItem>();
        static TypeMap()
        {
            IList<ClassItem> ClassList = new List<ClassItem>();
            ClassList.Add(new ClassItem(1,typeof(DBNull)));
            ClassList.Add(new ClassItem(2,typeof(object)));
            ClassList.Add(new ClassItem(3,typeof(bool)));
            ClassList.Add(new ClassItem(4,typeof(char)));
            ClassList.Add(new ClassItem(5,typeof(byte),typeof(SByte)));
            ClassList.Add(new ClassItem(7,typeof(short),typeof(UInt16)));
            ClassList.Add(new ClassItem(9,typeof(int),typeof(UInt32)));
            ClassList.Add(new ClassItem(11,typeof(long),typeof(UInt64)));
            ClassList.Add(new ClassItem(13,typeof(float)));
            ClassList.Add(new ClassItem(14,typeof(double)));
            ClassList.Add(new ClassItem(15,typeof(decimal)));
            ClassList.Add(new ClassItem(16,typeof(DateTime)));
            ClassList.Add(new ClassItem(18,typeof(string)));
            ClassList.Add(new ClassItem(19,typeof(IList)));
            ClassList.Add(new ClassItem(22,typeof(KeyValuePair<,>)));
            ClassList.Add(new ClassItem(23,typeof(Array)));
            ClassList.Add(new ClassItem(24,typeof(IDictionary)));
            foreach (ClassItem item in ClassList)
            {
                uint id = item.TypeId;
                foreach(Type t in item.Types)
                {
                    TypeIdMap[t] = item;
                }
                IdTypeMap[id] = item;
            }
            LoadCustemType();
        }

        private static void LoadCustemType()
        {
            var assemblies = AppDomain.CurrentDomain.GetAssemblies();
            foreach (var assembly in assemblies)
            {
                Module[] modules = null;
                try
                {
                    modules = assembly.GetModules();
                }
                catch
                {
                    continue;
                }
                foreach (var module in modules)
                {
                    Type[] types = null;
                    try
                    {
                        types = module.GetTypes();
                    }
                    catch
                    {
                        continue;
                    }
                    foreach (var type in types)
                    {
                        try
                        {
                            var attrs = type.GetCustomAttributes(typeof(GaeaSerializableAttribute), false);
                            if (attrs.Length > 0)
                            {
                                uint typeId = type.GetTypeId();
                                var ci = new ClassItem(typeId,type);
                                IdTypeMap[typeId] = ci;
                                TypeIdMap[type] = ci;
                            }
                        }
                        catch
                        {
                        }
                    }
                }
            }
            string[] files = Directory.GetFiles(AppDomain.CurrentDomain.BaseDirectory, "*.dll", SearchOption.AllDirectories);
            foreach (var file in files)
            {
                try
                {
                    FileInfo fi = new FileInfo(file);
                    var modules = System.Reflection.Assembly.Load(fi.Name.Replace(fi.Extension, string.Empty)).GetModules();
                    foreach (var module in modules)
                    {
                        try
                        {
                            var types = module.GetTypes();
                            foreach (var type in types)
                            {
                                var attrs = type.GetCustomAttributes(typeof(GaeaSerializableAttribute), false);
                                if (attrs.Length > 0)
                                {
                                    uint typeId = type.GetTypeId();
                                    var ci = new ClassItem(typeId,type);
                                    IdTypeMap[typeId] = ci;
                                    TypeIdMap[type] = ci;
                                }
                            }
                        }
                        catch { }
                    }
                }
                catch
                { }
            }
        }

        public static uint GetTypeId(Type type)
        {
            if (type.IsGenericType)
            {
                type = type.GetGenericTypeDefinition();
            }
            else if (type.IsArray)
            {
                type = typeof(Array);
            }
            if (typeof(IDictionary).IsAssignableFrom(type))
            {
                type = typeof(IDictionary);
            }
            else if (typeof(IList).IsAssignableFrom(type) && type!=typeof(Array))
            {
                type = typeof(IList);
            }
            ClassItem ci;
            if (TypeIdMap.TryGetValue(type, out ci))
            {
                return ci.TypeId;
            }
            var attributes = type.GetCustomAttributes(typeof(GaeaSerializableAttribute), false);
            if (attributes.Length > 0)
            {
                string name = ((GaeaSerializableAttribute)attributes[0]).Name;
                if (string.IsNullOrEmpty(name))
                {
                    name = type.Name;
                }
                uint typeId = (uint)name.GaeaHashCode();
                var ci2 = new ClassItem(typeId,type);
                IdTypeMap[typeId] = ci;
                TypeIdMap[type] = ci;
                return typeId;
            }
            throw new DisallowedSerializeException(type);
        }
        public static Type GetType(uint typeId)
        {
            ClassItem ci;
            if (IdTypeMap.TryGetValue(typeId, out ci))
            {
                return ci.Types[0];
            }
            return null;
        }
    }

    public class ClassItem
    {
        private Type[] _types;
        private uint _typeId;
        public ClassItem(uint typeid, params Type[] types)
        {
            _types = types;
            _typeId = typeid;
        }

        public Type[] Types
        {
            get { return _types; }
        }
        public uint TypeId
        {
            get { return _typeId; }
        }
    }
}

File: serializer/src/com/bj58/spat/gaea/serializer/component/TypeMap.java
/*
 *  Copyright Beijing 58 Information Technology Co.,Ltd.
 *
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package com.bj58.spat.gaea.serializer.component;

import java.util.Map;
import java.util.Set;
import java.util.Date;
import java.util.List;
import java.util.HashMap;
import java.util.ArrayList;
import java.io.IOException;
import java.math.BigDecimal;
import java.lang.reflect.Array;
import java.net.URISyntaxException;
import java.net.MalformedURLException;

import com.bj58.spat.gaea.serializer.classes.DBNull;
import com.bj58.spat.gaea.serializer.classes.GKeyValuePair;
import com.bj58.spat.gaea.serializer.component.annotation.GaeaSerializable;
import com.bj58.spat.gaea.serializer.component.exception.DisallowedSerializeException;
import com.bj58.spat.gaea.serializer.component.helper.StrHelper;
import com.bj58.spat.gaea.serializer.component.ClassItem;

/**
 * TypeMap
 * 
 * @author Service Platform Architecture Team (spat@58.com)
 */
public final class TypeMap {

    private static Map<Class, ClassItem> TypeIdMap = new HashMap<Class, ClassItem>();
    private static Map<Integer, ClassItem> IdTypeMap = new HashMap<Integer, ClassItem>();

    public static void InitTypeMap() {
        TypeIdMap.clear();
        IdTypeMap.clear();
        ArrayList<ClassItem> ClassList = new ArrayList<ClassItem>();
        ClassList.add(new ClassItem(1, DBNull.class));
        ClassList.add(new ClassItem(2, Object.class));
        ClassList.add(new ClassItem(3, Boolean.class, boolean.class));
        ClassList.add(new ClassItem(4, Character.class, char.class));
        ClassList.add(new ClassItem(5, Byte.class, byte.class));
        ClassList.add(new ClassItem(7, Short.class, short.class));
        ClassList.add(new ClassItem(9, Integer.class, int.class));
        ClassList.add(new ClassItem(11, Long.class, long.class));
        ClassList.add(new ClassItem(13, Float.class, float.class));
        ClassList.add(new ClassItem(14, Double.class, double.class));
        ClassList.add(new ClassItem(15, BigDecimal.class));
        ClassList.add(new ClassItem(16, Date.class, java.sql.Date.class, java.sql.Time.class, java.sql.Timestamp.class));
        ClassList.add(new ClassItem(18, String.class));
        ClassList.add(new ClassItem(19, List.class));
        ClassList.add(new ClassItem(22, GKeyValuePair.class));
        ClassList.add(new ClassItem(23, Array.class));
        ClassList.add(new ClassItem(24, Map.class));
        for (ClassItem item : ClassList) {
            int id = item.getTypeId();
            Class[] types = item.getTypes();
            for (Class c : types) {
                TypeIdMap.put(c, item);
            }
            IdTypeMap.put(id, item);
        }

        String scanType = System.getProperty("gaea.serializer.scantype");
        if (scanType != null && scanType.equals("asyn")) {
            Thread th = new Thread(new Runnable() {
                @Override
                public void run() {
                    System.out.println("Scan jar files begin!");
                    try {
                        LoadCustmeType();
                    } catch (Exception ex) {
                        ex.printStackTrace();
                    }
                    System.out.println("Scan jar files completed!");
                }
            });
            th.start();
        } else {
            System.out.println("Scan jar files begin!");
            try {
                LoadCustmeType();
            } catch (Exception ex) {
                ex.printStackTrace();
            }
            System.out.println("Scan jar files completed!");
        }
    }

    private static void LoadCustmeType() throws URISyntaxException, IOException, MalformedURLException, ClassNotFoundException {
        ClassScaner cs = new ClassScaner();
        String basePakage = System.getProperty("gaea.serializer.basepakage");
        if (basePakage == null) {
            basePakage = StrHelper.EmptyString;
        }
        Set<Class> classes = cs.scan(basePakage.split(";"));
        for (Class c : classes) {
            System.out.println("scaning " + c.getPackage().getName() + "." + c.getName());
            try {
                GaeaSerializable ann = (GaeaSerializable) c.getAnnotation(GaeaSerializable.class);
                if (ann != null) {
                    String name = ann.name();
                    if (name.equals(StrHelper.EmptyString)) {
                        name = c.getSimpleName();
                    }
                    int typeId = StrHelper.GetHashcode(name);
                    TypeIdMap.put(c, new ClassItem(typeId, c));
                    IdTypeMap.put(typeId, new ClassItem(typeId, c));
                }
            } catch (Exception ex) {
                ex.printStackTrace();
            }
        }
    }

    public static Class getClass(int typeId) {
        /*****************兼容之前版本***********************/
        if(typeId==6){
            typeId =5;
        }else if(typeId==20 || typeId == 21){
            typeId = 19;
        }else if(typeId == 25){
            typeId = 24;
        }
        /**************************************************/
        ClassItem ci = IdTypeMap.get(typeId);
        if (ci != null) {
            return ci.getType();
        }
        return null;
    }

    public static int getTypeId(Class type) throws DisallowedSerializeException {
        int typeId = 0;
        if (type.isArray()) {
            type = Array.class;
        } else if (Map.class.isAssignableFrom(type)) {
            type = Map.class;
        } else if (List.class.isAssignableFrom(type)) {
            type = List.class;
        }
        ClassItem ci = TypeIdMap.get(type);
        if (ci != null) {
            typeId = ci.getTypeId();
        } else {
            GaeaSerializable ann = (GaeaSerializable) type.getAnnotation(GaeaSerializable.class);
            if (ann == null) {
                throw new DisallowedSerializeException(type);
            }
            String name = ann.name();
            if (name.equals(StrHelper.EmptyString)) {
                name = type.getSimpleName();
            }
            typeId = StrHelper.GetHashcode(name);
            setTypeMap(type, typeId);
        }
        return typeId;
    }

    public static void setTypeMap(Class type, int typeId) {
        ClassItem ci = new ClassItem(typeId, type);
        TypeIdMap.put(type, ci);
        IdTypeMap.put(typeId, ci);
    }
}

class ClassItem {

    private Class[] Types;
    private int TypeId;

    public ClassItem(int typeids, Class... types) {
        Types = types;
        TypeId = typeids;
    }

    public Class getType() {
        return Types[0];
    }

    public Class[] getTypes() {
        return Types;
    }

    public int getTypeId() {
        return TypeId;
    }
}
